File: LeetCode_Algorithm/994(Rotting Oranges)/994.py
from typing import List
class Solution:
    def __init__(self):
        self.minute = 0
        self.dr = [0,1,0,-1]
        self.dc = [1,0,-1,0]
        self.rotten_orange = []
        self.fresh_orange = 0
    def orangesRotting(self, grid: List[List[int]]) -> int:
        # O(n) O(1) 97ms 26%
        # 먼저 오렌지 개수와 썩은 오렌지 위치를 계산
        for r in range(len(grid)):
            for c in range(len(grid[0])):
                if grid[r][c] == 1:
                    self.fresh_orange += 1
                elif grid[r][c] == 2:
                    self.rotten_orange.append([r,c])
        while True:
            temp = []
            while self.rotten_orange:
                r,c = self.rotten_orange.pop(0)
                for i in range(4):
                    new_r = r + self.dr[i]; new_c = c + self.dc[i]
                    if 0 <= new_r < len(grid) and 0 <= new_c < len(grid[0]):
                        if grid[new_r][new_c] == 1:
                            self.fresh_orange -= 1;
                            grid[new_r][new_c] = 2; temp.append([new_r,new_c])
            self.minute += 1
            if not temp:
                break
            self.rotten_orange = temp
        if self.fresh_orange > 0: return -1
        return self.minute-1 # 가장 마지막에 썩는 과일 체크 시간 제거
            


print(Solution().orangesRotting([[2,1,1],[1,1,0],[0,1,1]]))
print(Solution().orangesRotting([[1,2]]))

File: LeetCode_Algorithm/994(Rotting Oranges)/994.js
var orangesRotting = function(grid) {
  // O(n) O(1) 426ms 5%
  let freshOrange = 0;
  let minute = 0;
  let rotten_location = []
  const dr = [0,-1,0,1]; const dc = [1,0,-1,0]
  for (let r in grid) {
    for (let c in grid[0]){
      if (grid[r][c] == 1){
        freshOrange += 1
      } else if (grid[r][c] == 2) {
        rotten_location.push([r,c])
      }
    }
  }
  while (true) {
    let temp = []
    while (rotten_location.length > 0) {
      [r,c] = rotten_location.pop()
      for (let i=0; i < 4; i++) {
        let newR = parseInt(r) + dr[i]; let newC = parseInt(c) + dc[i]
        console.log(newR, newC)
        if (-1 < newR && newR < grid.length && -1 < newC && newC < grid[0].length) {
          if (grid[newR][newC] == 1) {
            console.log(newR, newC)
            freshOrange -= 1; grid[newR][newC] = 2; temp.push([newR,newC])
          }
        }
      }
    }
    minute += 1
    // console.log(temp, minute)
    if (temp.length == 0) {
      break;
    }
    rotten_location = temp
  }
  if (freshOrange > 0) {
    return -1
  }
  return minute-1
};

// console.log(orangesRotting([[2,1,1],[1,1,0],[0,1,1]]))
console.log(orangesRotting([[1,2]]))
